### DISCUSSION TITLE
[2023-02-13] Performance Team Demo

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#5642] [Weekly Demo Day]
Let me start! :) 

I was working on improving the Electron integration, this involves:
- Unifying how we pass flags between Electron, Rust, and the code running in the browser.
- Updating Electron to the newest version.
- Reviewing possible Electron/Chrome flags to make it more performant and working better on different systems. 
- The whole electron management code was rewritten from JS to TS. It was ugly because it was first a test and it survived till now. Now it's properly written and documented. 

The PR is opened, still not all code cleaned, will be cleaned for tomorrow: https://github.com/enso-org/enso/pull/5633

As a result, we've got Electron app working on M2 Macs (not crashing anymore, I hope). Also, it should remove some of community issues, including this one: https://github.com/enso-org/enso/issues/3801

As a side improvement, the command line interface is way nicer now, here is the screenshot of running the electron app with `-help` command (not all switches are visible, as we have almost a 100 switches now, including chrome/electron switches to debug the app).

<img width="845" alt="image" src="https://user-images.githubusercontent.com/1623053/218465786-1b188f06-69f0-4b07-bc11-de3ac75cd495.png">

**Comment by JaroslavTulach** [comment]
Based on the [Feb 6 Demo review feedback](https://docs.google.com/document/d/1-jG7HQoPY7KASG8NZ1PUDzqqd5ZL7wMzVSSMCcNFWMI/)  individual issues were reported and are being worked on:

- https://github.com/enso-org/enso/issues/5068 (Dmitry, Jaroslav)
- https://github.com/enso-org/enso/issues/5568 (Hubert)
    - https://github.com/enso-org/enso/issues/5567 (Hubert)

Low priority tasks reported:

- https://github.com/enso-org/enso/issues/5569
- https://github.com/enso-org/enso/issues/5570

Custom [Hashing](https://github.com/enso-org/enso/pull/4067) has been fully integrated :
- speeds `Vector.distinct` up
- supports custom hashcode and equals implementations 
- consistency guaranteed
- support for inequality operators (e.g. `<`, `<=`, `>`, `>=`)
- no more incompatible API changes planned

We can move forward and bugfix and [speed up](https://github.com/enso-org/enso/issues/5636).

**Comment by kazcw** [comment]
Last week, the frontend parser transition landed (#4113).

This is what startup looked like before (opening "Orders", zoomed to the part after we receive project source code from the LS):
![image](https://user-images.githubusercontent.com/1047859/218505670-90f64ba8-80a6-48cb-a715-04d1831f8a28.png)

The circled portion is over 500ms of CPU work in `parse_with_metadata`; that has been eliminated. The new `parse_with_metadata` takes 1.8ms.

Since then I've been working on optimizing shader compilation scheduling (#5606). I'll finish that today. Next I'll profile some objectives #5603 and look for the next optimization.

**Comment by farmaazon** [comment]
Spent much time on ~~taming Github Projects~~ removing all artifacts and improving the rendering of the shapes rendered from the texture. Tested on the real icon from Component Browser - that showed a few new things. Now, I finally made the cached icon (up) looking almost exactly as rendered from traditional shape system (bottom) _if they are aligned to the pixels_:
![image](https://user-images.githubusercontent.com/3919101/218506510-634077ea-9f33-40f1-b70d-be8698e6e0af.png)

Still, things are getting worse when _they are not_:
![image](https://user-images.githubusercontent.com/3919101/218507479-d47f3b12-e070-425a-ac47-6b7b6f9c36c3.png)

Tried to fix that, but this is extremely not easy. However, after a talk with @wdanilo we decided, that the normal shapes, if not aligned to pixels, are bad anyway, and we rather aim to snap things to pixels if they are not moving.

**Comment by mwu-tow** [comment]
* Reworked watch command to use Electron exutable directly. https://github.com/enso-org/enso/pull/5634
* * New „ide watch”;
* * New „ide start”.
* Related refactorings.
* Debugging CI issues:
* * Issues with Engine builds — fixed by Hubert. https://github.com/enso-org/enso/pull/5630
* Plans:
* * Finishing PR. https://github.com/enso-org/enso/pull/5634
* * Reviewing and testing Wojciech's https://github.com/enso-org/enso/pull/5633

**Comment by Frizi** [comment]
Refined argument placeholders and widget support for more complicated expression types.

Added support for widgets in infix expressions (right now only used for file paths)
<img width="306" alt="image" src="https://user-images.githubusercontent.com/919491/218736181-98965cd4-2a8e-4f7a-bbf8-ab302ac5b22c.png">

Widgets and placeholders are handled for all chained methods within a node
<img width="292" alt="image" src="https://user-images.githubusercontent.com/919491/218736249-a0190115-623e-4e66-aff4-90eb2a50685d.png">

The qualified method call convention and static methods no longer confuse the argument placeholders
<img width="374" alt="image" src="https://user-images.githubusercontent.com/919491/218736628-24073f5c-0512-4b37-a271-0248bc954802.png">

Type constructor expressions now receive placeholder arguments. The placeholders work on nested expressions within a node.
<img width="405" alt="image" src="https://user-images.githubusercontent.com/919491/218737379-b53ff125-3910-48f3-bb9f-a0e7b1684ab9.png">
